accord-core/src/test/java/accord/impl/list/ListAgent.java

```diff
@@ -46,11 +52,16 @@ public ListAgent(long timeout, Consumer<Throwable> onFailure, Consumer<Runnable>
     @Override
     public void onRecover(Node node, Result success, Throwable fail)
     {
+        if (fail != null)
+        {
+            checkState(success == null, "fail (%s) and success (%s) are both not null", fail, success);
+            // We don't really process errors for Recover here even though it is provided in the interface
```

Comment:
we should report back still, else the future can hang forever right?

Reply:
We can't report back because without the `success` there is no `requestId` to respond with. We can't use uncaught exception handler because the exception isn't typically uncaught tit's actually been passed by the caller of `onRecover`. `onRecover` is an odd one because it's kind of informative, but it isn't the callback for recover it's just some notification the agent gets.

Reply:
We can't report back because without the `success` there is no `requestId` to respond with. We can't use uncaught exception handler because the exception isn't typically uncaught tit's actually been passed by to the callback of `Recover`. `onRecover` is an odd one because it's kind of informative, but it isn't the callback for recover it's just some notification the agent gets.

I need to look at why `onRecover` is involved in sending the reply at all. Why doesn't it get done in the callback for `Recover`? Why does it exist at all?

Reply:
This is only for the ListAgent, which is used only for testing purposes. We keep enough information to issue a response directly to clients for testing so we get more responses to analyse the correctness of.